From 02bf82a5bc41dd764afc4054ce45cf07ebd87704 Mon Sep 17 00:00:00 2001
From: Alexis Williams <alexis@typedr.at>
Date: Tue, 29 Sep 2026 19:05:00 -0700
Subject: [PATCH 19/24] Call more callbacks through their own function types

Calling a function through a pointer to a type with a different
parameter list or return type is undefined behaviour; WebAssembly
checks the type of every indirect call and traps with "function
signature mismatch". These sites were missed by the earlier fixes:

- gtypemodule: GTypeModule's class_init took one argument, so every
  GIO extension point (and any GTypeModule) trapped on first use.
- gbookmarkfile: g_bookmark_file_copy() passed g_strdup() and one-
  argument copy functions to g_list_copy_deep() as GCopyFunc.
- gsourceclosure: g_closure_ref(), which returns the closure, was the
  void-returning ref function of a closure source's callback funcs.
- gdbusinterfaceskeleton, gdbusobjectmanagerserver, gdbusobjectproxy,
  gdbusobjectskeleton, gmemorymonitorbase, and the interface init
  functions gdbus-codegen emits: one-argument GInterfaceInitFunc.
- gtester: g_free() passed to g_slist_foreach() as a GFunc.

Upstream-Status: Inappropriate [Wasm-specific]
---
 gio/gdbus-2.0/codegen/codegen.py | 22 ++++++++++------------
 gio/gdbusinterfaceskeleton.c     |  6 ++++--
 gio/gdbusobjectmanagerserver.c   |  6 ++++--
 gio/gdbusobjectproxy.c           |  6 ++++--
 gio/gdbusobjectskeleton.c        |  6 ++++--
 gio/gmemorymonitorbase.c         | 12 ++++++++----
 glib/gbookmarkfile.c             | 29 ++++++++++++++++++++++++++---
 glib/gtester.c                   |  3 +--
 gobject/gsourceclosure.c         | 11 ++++++++++-
 gobject/gtypemodule.c            |  3 ++-
 10 files changed, 73 insertions(+), 31 deletions(-)

diff --git a/gio/gdbus-2.0/codegen/codegen.py b/gio/gdbus-2.0/codegen/codegen.py
index 2add4c810..fd2ac2ed9 100644
--- a/gio/gdbus-2.0/codegen/codegen.py
+++ b/gio/gdbus-2.0/codegen/codegen.py
@@ -3148,7 +3148,7 @@ class CodeGenerator:
         )
 
         self.outfile.write(
-            "static void %s_proxy_iface_init (%sIface *iface);\n"
+            "static void %s_proxy_iface_init (%sIface *iface, gpointer iface_data);\n"
             "\n" % (i.name_lower, i.camel_name)
         )
         self.outfile.write("#if GLIB_VERSION_MAX_ALLOWED >= GLIB_VERSION_2_38\n")
@@ -3526,9 +3526,8 @@ class CodeGenerator:
             "%s_proxy_iface_init (%sIface *iface" % (i.name_lower, i.camel_name)
         )
         if len(i.properties) == 0:
-            self.outfile.write(" G_GNUC_UNUSED)\n")
-        else:
-            self.outfile.write(")\n")
+            self.outfile.write(" G_GNUC_UNUSED")
+        self.outfile.write(", gpointer iface_data G_GNUC_UNUSED)\n")
         self.outfile.write("{\n")
         for p in i.properties:
             self.outfile.write(
@@ -4135,7 +4134,7 @@ class CodeGenerator:
             self.outfile.write("}\n" "\n")
 
         self.outfile.write(
-            "static void %s_skeleton_iface_init (%sIface *iface);\n"
+            "static void %s_skeleton_iface_init (%sIface *iface, gpointer iface_data);\n"
             % (i.name_lower, i.camel_name)
         )
 
@@ -4530,9 +4529,8 @@ class CodeGenerator:
             "%s_skeleton_iface_init (%sIface *iface" % (i.name_lower, i.camel_name)
         )
         if len(i.signals) == 0 and len(i.properties) == 0:
-            self.outfile.write(" G_GNUC_UNUSED)\n")
-        else:
-            self.outfile.write(")\n")
+            self.outfile.write(" G_GNUC_UNUSED")
+        self.outfile.write(", gpointer iface_data G_GNUC_UNUSED)\n")
         self.outfile.write("{\n")
         for s in i.signals:
             self.outfile.write(
@@ -4799,14 +4797,14 @@ class CodeGenerator:
         # class boilerplate
         self.outfile.write(
             "static void\n"
-            "%sobject_proxy__%sobject_iface_init (%sObjectIface *iface G_GNUC_UNUSED)\n"
+            "%sobject_proxy__%sobject_iface_init (%sObjectIface *iface G_GNUC_UNUSED, gpointer iface_data G_GNUC_UNUSED)\n"
             "{\n"
             "}\n"
             "\n" % (self.ns_lower, self.ns_lower, self.namespace)
         )
         self.outfile.write(
             "static void\n"
-            "%sobject_proxy__g_dbus_object_iface_init (GDBusObjectIface *iface)\n"
+            "%sobject_proxy__g_dbus_object_iface_init (GDBusObjectIface *iface, gpointer iface_data G_GNUC_UNUSED)\n"
             "{\n"
             "  iface->interface_added = %sobject_notify;\n"
             "  iface->interface_removed = %sobject_notify;\n"
@@ -4951,7 +4949,7 @@ class CodeGenerator:
         # class boilerplate
         self.outfile.write(
             "static void\n"
-            "%sobject_skeleton__%sobject_iface_init (%sObjectIface *iface G_GNUC_UNUSED)\n"
+            "%sobject_skeleton__%sobject_iface_init (%sObjectIface *iface G_GNUC_UNUSED, gpointer iface_data G_GNUC_UNUSED)\n"
             "{\n"
             "}\n"
             "\n" % (self.ns_lower, self.ns_lower, self.namespace)
@@ -4959,7 +4957,7 @@ class CodeGenerator:
         self.outfile.write("\n")
         self.outfile.write(
             "static void\n"
-            "%sobject_skeleton__g_dbus_object_iface_init (GDBusObjectIface *iface)\n"
+            "%sobject_skeleton__g_dbus_object_iface_init (GDBusObjectIface *iface, gpointer iface_data G_GNUC_UNUSED)\n"
             "{\n"
             "  iface->interface_added = %sobject_notify;\n"
             "  iface->interface_removed = %sobject_notify;\n"
diff --git a/gio/gdbusinterfaceskeleton.c b/gio/gdbusinterfaceskeleton.c
index ba4430259..34bec28b3 100644
--- a/gio/gdbusinterfaceskeleton.c
+++ b/gio/gdbusinterfaceskeleton.c
@@ -75,7 +75,8 @@ enum
 
 static guint signals[LAST_SIGNAL] = {0};
 
-static void     dbus_interface_interface_init                      (GDBusInterfaceIface    *iface);
+static void     dbus_interface_interface_init                      (GDBusInterfaceIface    *iface,
+                                                                    gpointer                iface_data);
 
 static void     set_object_path_locked                             (GDBusInterfaceSkeleton *interface_,
                                                                     const gchar            *object_path);
@@ -451,7 +452,8 @@ g_dbus_interface_skeleton_set_object (GDBusInterface *interface_,
 }
 
 static void
-dbus_interface_interface_init (GDBusInterfaceIface *iface)
+dbus_interface_interface_init (GDBusInterfaceIface *iface,
+                               gpointer             iface_data)
 {
   iface->get_info    = _g_dbus_interface_skeleton_get_info;
   iface->get_object  = g_dbus_interface_skeleton_get_object;
diff --git a/gio/gdbusobjectmanagerserver.c b/gio/gdbusobjectmanagerserver.c
index aec199fbb..10f0ffea3 100644
--- a/gio/gdbusobjectmanagerserver.c
+++ b/gio/gdbusobjectmanagerserver.c
@@ -108,7 +108,8 @@ enum
   PROP_OBJECT_PATH
 };
 
-static void dbus_object_manager_interface_init (GDBusObjectManagerIface *iface);
+static void dbus_object_manager_interface_init (GDBusObjectManagerIface *iface,
+                                                gpointer                 iface_data);
 
 G_DEFINE_TYPE_WITH_CODE (GDBusObjectManagerServer, g_dbus_object_manager_server, G_TYPE_OBJECT,
                          G_ADD_PRIVATE (GDBusObjectManagerServer)
@@ -1088,7 +1089,8 @@ g_dbus_object_manager_server_get_interface  (GDBusObjectManager  *_manager,
 }
 
 static void
-dbus_object_manager_interface_init (GDBusObjectManagerIface *iface)
+dbus_object_manager_interface_init (GDBusObjectManagerIface *iface,
+                                    gpointer                 iface_data)
 {
   iface->get_object_path = g_dbus_object_manager_server_get_object_path;
   iface->get_objects     = g_dbus_object_manager_server_get_objects;
diff --git a/gio/gdbusobjectproxy.c b/gio/gdbusobjectproxy.c
index 3ea7e440a..dd7b953f4 100644
--- a/gio/gdbusobjectproxy.c
+++ b/gio/gdbusobjectproxy.c
@@ -57,7 +57,8 @@ enum
   PROP_G_CONNECTION
 };
 
-static void dbus_object_interface_init (GDBusObjectIface *iface);
+static void dbus_object_interface_init (GDBusObjectIface *iface,
+                                        gpointer          iface_data);
 
 G_DEFINE_TYPE_WITH_CODE (GDBusObjectProxy, g_dbus_object_proxy, G_TYPE_OBJECT,
                          G_ADD_PRIVATE (GDBusObjectProxy)
@@ -345,7 +346,8 @@ _g_dbus_object_proxy_remove_interface (GDBusObjectProxy *proxy,
 }
 
 static void
-dbus_object_interface_init (GDBusObjectIface *iface)
+dbus_object_interface_init (GDBusObjectIface *iface,
+                            gpointer          iface_data)
 {
   iface->get_object_path       = g_dbus_object_proxy_get_object_path;
   iface->get_interfaces        = g_dbus_object_proxy_get_interfaces;
diff --git a/gio/gdbusobjectskeleton.c b/gio/gdbusobjectskeleton.c
index 066d332e9..6b79838cd 100644
--- a/gio/gdbusobjectskeleton.c
+++ b/gio/gdbusobjectskeleton.c
@@ -66,7 +66,8 @@ enum
 
 static guint signals[LAST_SIGNAL] = {0};
 
-static void dbus_object_interface_init (GDBusObjectIface *iface);
+static void dbus_object_interface_init (GDBusObjectIface *iface,
+                                        gpointer          iface_data);
 
 G_DEFINE_TYPE_WITH_CODE (GDBusObjectSkeleton, g_dbus_object_skeleton, G_TYPE_OBJECT,
                          G_ADD_PRIVATE (GDBusObjectSkeleton)
@@ -491,7 +492,8 @@ g_dbus_object_skeleton_flush (GDBusObjectSkeleton *object)
 }
 
 static void
-dbus_object_interface_init (GDBusObjectIface *iface)
+dbus_object_interface_init (GDBusObjectIface *iface,
+                            gpointer          iface_data)
 {
   iface->get_object_path = g_dbus_object_skeleton_get_object_path;
   iface->get_interfaces  = g_dbus_object_skeleton_get_interfaces;
diff --git a/gio/gmemorymonitorbase.c b/gio/gmemorymonitorbase.c
index 56b083f7e..b0a3aa7ec 100644
--- a/gio/gmemorymonitorbase.c
+++ b/gio/gmemorymonitorbase.c
@@ -50,8 +50,10 @@
 
 #define G_MEMORY_MONITOR_BASE_GET_INITABLE_IFACE(o) (G_TYPE_INSTANCE_GET_INTERFACE ((o), G_TYPE_INITABLE, GInitable))
 
-static void g_memory_monitor_base_iface_init (GMemoryMonitorInterface *iface);
-static void g_memory_monitor_base_initable_iface_init (GInitableIface *iface);
+static void g_memory_monitor_base_iface_init (GMemoryMonitorInterface *iface,
+                                              gpointer                 iface_data);
+static void g_memory_monitor_base_initable_iface_init (GInitableIface *iface,
+                                                       gpointer        iface_data);
 
 typedef struct
 {
@@ -190,12 +192,14 @@ g_memory_monitor_base_class_init (GMemoryMonitorBaseClass *klass)
 }
 
 static void
-g_memory_monitor_base_iface_init (GMemoryMonitorInterface *monitor_iface)
+g_memory_monitor_base_iface_init (GMemoryMonitorInterface *monitor_iface,
+                                  gpointer                 iface_data)
 {
 }
 
 static void
-g_memory_monitor_base_initable_iface_init (GInitableIface *iface)
+g_memory_monitor_base_initable_iface_init (GInitableIface *iface,
+                                           gpointer        iface_data)
 {
   iface->init = g_memory_monitor_base_initable_init;
 }
diff --git a/glib/gbookmarkfile.c b/glib/gbookmarkfile.c
index 7392b1e07..6c344e4e3 100644
--- a/glib/gbookmarkfile.c
+++ b/glib/gbookmarkfile.c
@@ -355,6 +355,22 @@ bookmark_metadata_free (BookmarkMetadata *metadata)
   g_slice_free (BookmarkMetadata, metadata);
 }
 
+/* The copy functions as #GCopyFunc: calling them through that
+ * two-argument type is undefined behaviour, and traps on WebAssembly. */
+static gpointer
+copy_string (gconstpointer src,
+             gpointer      data)
+{
+  return g_strdup (src);
+}
+
+static gpointer
+copy_app_info (gconstpointer src,
+               gpointer      data)
+{
+  return bookmark_app_info_copy ((BookmarkAppInfo *) src);
+}
+
 static BookmarkMetadata *
 bookmark_metadata_copy (BookmarkMetadata *metadata)
 {
@@ -370,9 +386,9 @@ bookmark_metadata_copy (BookmarkMetadata *metadata)
   copy->icon_href = g_strdup (metadata->icon_href);
   copy->icon_mime = g_strdup (metadata->icon_mime);
 
-  copy->groups = g_list_copy_deep (metadata->groups, (GCopyFunc) g_strdup, NULL);
+  copy->groups = g_list_copy_deep (metadata->groups, copy_string, NULL);
   copy->applications =
-    g_list_copy_deep (metadata->applications, (GCopyFunc) bookmark_app_info_copy, NULL);
+    g_list_copy_deep (metadata->applications, copy_app_info, NULL);
 
   for (l = copy->applications; l; l = l->next)
     {
@@ -1716,6 +1732,13 @@ g_bookmark_file_new (void)
   return bookmark;
 }
 
+static gpointer
+copy_item (gconstpointer src,
+           gpointer      data)
+{
+  return bookmark_item_copy ((BookmarkItem *) src);
+}
+
 /**
  * g_bookmark_file_copy:
  * @bookmark: A #GBookmarkFile
@@ -1738,7 +1761,7 @@ g_bookmark_file_copy (GBookmarkFile *bookmark)
   copy = g_bookmark_file_new ();
   copy->title = g_strdup (bookmark->title);
   copy->description = g_strdup (bookmark->description);
-  copy->items = g_list_copy_deep (bookmark->items, (GCopyFunc) bookmark_item_copy, NULL);
+  copy->items = g_list_copy_deep (bookmark->items, copy_item, NULL);
 
   for (l = copy->items; l; l = l->next)
     {
diff --git a/glib/gtester.c b/glib/gtester.c
index 40b4706f9..256115231 100644
--- a/glib/gtester.c
+++ b/glib/gtester.c
@@ -395,8 +395,7 @@ launch_test_binary (const char *binary,
                             NULL,       /* standard_output */
                             NULL,       /* standard_error */
                             &error);
-  g_slist_foreach (free_list, (void(*)(void*,void*)) g_free, NULL);
-  g_slist_free (free_list);
+  g_slist_free_full (free_list, g_free);
   free_list = NULL;
   close (report_pipe[1]);
 
diff --git a/gobject/gsourceclosure.c b/gobject/gsourceclosure.c
index f4076abdc..c5718d935 100644
--- a/gobject/gsourceclosure.c
+++ b/gobject/gsourceclosure.c
@@ -222,8 +222,17 @@ closure_callback_get (gpointer     cb_data,
   *data = cb_data;
 }
 
+/* g_closure_ref() returns the closure, so calling it through the
+ * void-returning ref type is undefined behaviour, and traps on
+ * WebAssembly. */
+static void
+closure_callback_ref (gpointer cb_data)
+{
+  g_closure_ref (cb_data);
+}
+
 static GSourceCallbackFuncs closure_callback_funcs = {
-  (void (*) (gpointer)) g_closure_ref,
+  closure_callback_ref,
   (void (*) (gpointer)) g_closure_unref,
   closure_callback_get
 };
diff --git a/gobject/gtypemodule.c b/gobject/gtypemodule.c
index f06f521bc..c6a5ef6b4 100644
--- a/gobject/gtypemodule.c
+++ b/gobject/gtypemodule.c
@@ -123,7 +123,8 @@ g_type_module_finalize (GObject *object)
 }
 
 static void
-g_type_module_class_init (GTypeModuleClass *class)
+g_type_module_class_init (GTypeModuleClass *class,
+                          gpointer          class_data)
 {
   GObjectClass *gobject_class = G_OBJECT_CLASS (class);
 
